smarts/core/sensors.py

```diff
@@ -1209,3 +1239,91 @@ def closest_point_on_lane(position, lane_id):
 
     def teardown(self):
         pass
+
+
+class SignalsSensor(Sensor):
+    """Reports state of traffic signals (lights) in the lanes ahead of vehicle."""
+
+    def __init__(self, vehicle, road_map: RoadMap, lookahead: float):
+        self._logger = logging.getLogger(self.__class__.__name__)
+        self._vehicle = vehicle
+        self._road_map = road_map
+        self._lookahead = lookahead
+
+    @staticmethod
+    def _is_signal_type(feature: RoadMap.Feature) -> bool:
+        # XXX:  eventually if we add other types of dynamic features, we'll need to update this.
+        return (
+            feature.type == RoadMap.FeatureType.FIXED_LOC_SIGNAL or feature.is_dynamic
+        )
+
+    def __call__(
+        self,
+        lane: Optional[RoadMap.Lane],
+        lane_pos: RefLinePoint,
+        state,  # VehicleState
+        plan: Plan,
+        provider_state,  # ProviderState
+    ) -> List[SignalObservation]:
+        result = []
+        if not lane:
+            return result
+        upcoming_signals = []
+        half_len = 0.5 * state.dimensions.length
```

Comment:
Is this necessary? By the time the car is half-way past the signal, there is already consequence for violating or obeying the signal. Or the signal still affects the rules even though the vehicle has passed it.

Reply:
Yeah, it's probably not.  I was trying to be a bit "safe" with respect to the "sampling rate" (i.e., stepping rate) of the vehicle whereby it might move adjacent to the signal within one step if it's still moving too fast to stop on time.  I.e., I was trying to allow for an agent to not be so good at stopping at signals precisely.

But this might not be what someone who's training an agent to stop at signals would want.  So I'm fine with taking this out.   Will do, but will add a comment about this....